File: packages/server/src/cost.ts
/** Static cost, depth and field-count estimation. Spec: spec/06 §5, spec/02 §5. */
import { annotation, fieldsOf, isPageRef, type ArgDef, type OpDef, type RayfoldSchemaIR, type Shape, type ShapeItem, type TypeRef } from "@rayfold/schema";
import { defaultShape, isScalarLike } from "./views.ts";

export interface CostEstimate {
  cost: number;
  depth: number;
  fields: number;
}

const DEFAULT_FIRST = 20;
/** Largest page the runtime serves; any page size the cost model cannot trust counts as this. */
const MAX_FIRST = 200;

/**
 * cost = op.base + op.perItem × first + Σ over selected fields of (field.base + field.perItem × childFirst) × mult
 * where mult is the product of page sizes enclosing the field (fields inside `items` of a Page<T> count `first` times).
 * A field's base defaults to 1 when it returns objects (an entity, an object, a page, a list of them) and to 0 when
 * it returns scalars or enums: those come with the row that is already loaded. The perItem of a page, op or field,
 * defaults to 1, so every row a page can return is charged. `@cost` overrides each default.
 */
export function estimateCost(ir: RayfoldSchemaIR, op: OpDef, args: Record<string, unknown>, shape: Shape, vars: Record<string, unknown> = {}): CostEstimate {
  const c = annotation(op, "cost");
  const base = num(c?.args["base"]) ?? 1;
  const perItem = num(c?.args["perItem"]) ?? (isPageRef(op.returns) ? 1 : 0);
  const first = isPageRef(op.returns) ? pageFirst(args, op.args) : 1;
  const acc = { fields: 0, depth: 0 };
  const shapeCost = walk(ir, op.returns, shape, 1, first, 1, acc, vars);
  return { cost: Math.max(1, base + perItem * first + shapeCost), depth: acc.depth, fields: acc.fields };
}

/** Fields without a sub-shape that resolve to objects get that type's default view, exactly as the executor does. */
function walk(ir: RayfoldSchemaIR, t: TypeRef, shape: Shape, mult: number, itemsFirst: number, depth: number, acc: { fields: number; depth: number }, vars: Record<string, unknown>): number {
  acc.depth = Math.max(acc.depth, depth);
  const isPage = isPageRef(t) || (t.kind === "list" && isPageRef(t.of));
  let total = 0;
  const visit = (items: ShapeItem[], ref: TypeRef): void => {
    const fields = fieldsOf(ir, ref) ?? [];
    for (const it of items) {
      switch (it.kind) {
        case "field": {
          acc.fields++;
          const f = fields.find((x) => x.name === it.name);
          const fc = f ? annotation(f, "cost") : undefined;
          const childFirst = f && isPageRef(f.type) ? pageFirst(substituteVars(it.args ?? {}, vars), f.args) : 1;
          const defaultBase = f && isScalarLike(ir, f.type) ? 0 : 1;
          const defaultPerItem = f && isPageRef(f.type) ? 1 : 0;
          total += mult * ((num(fc?.args["base"]) ?? defaultBase) + (num(fc?.args["perItem"]) ?? defaultPerItem) * childFirst);
          if (f && !isScalarLike(ir, f.type)) {
            const sub = it.shape ?? defaultShape(ir, f.type);
            const childMult = isPage && it.name === "items" ? mult * itemsFirst : mult;
            total += walk(ir, f.type, sub, childMult, childFirst, depth + 1, acc, vars);
          }
          break;
        }
        case "spread": {
          const v = ir.views[`${it.type}.${it.view}`];
          if (v) visit(v.shape.items, ref);
          break;
        }
        case "on":
          visit(it.shape.items, { kind: "named", name: it.type, nullable: false });
          break;
        case "defer":
          visit(it.shape.items, ref);
          break;
      }
    }
  };
  visit(shape.items, t);
  return total;
}

/**
 * Page size for costing. The estimate runs on the request as sent, before coercion, so anything that is not a whole
 * number from 0 to MAX_FIRST (negative, fractional, huge, a $ref, a missing variable) counts as MAX_FIRST:
 * bad input can raise the estimate but never lower it.
 */
function pageFirst(args: Record<string, unknown>, defs: ArgDef[]): number {
  const page = args["page"];
  if (page !== undefined) {
    if (!page || typeof page !== "object" || Array.isArray(page)) return MAX_FIRST;
    if ("first" in page) return pageSize((page as Record<string, unknown>)["first"]);
  }
  if ("first" in args) return pageSize(args["first"]);
  const def = defs.find((a) => a.name === "page")?.default;
  if (def && typeof def === "object" && !Array.isArray(def) && "first" in def) return pageSize((def as Record<string, unknown>)["first"]);
  const firstDef = defs.find((a) => a.name === "first")?.default;
  if (firstDef !== undefined) return pageSize(firstDef);
  return DEFAULT_FIRST;
}

function pageSize(v: unknown): number {
  return typeof v === "number" && Number.isInteger(v) && v >= 0 ? Math.min(v, MAX_FIRST) : MAX_FIRST;
}

function num(v: unknown): number | undefined {
  return typeof v === "number" ? v : undefined;
}

function substituteVars(v: unknown, vars: Record<string, unknown>): Record<string, unknown> {
  const walkV = (x: unknown): unknown => {
    if (x === null || typeof x !== "object") return x;
    if (Array.isArray(x)) return x.map(walkV);
    const o = x as Record<string, unknown>;
    if (typeof o["$var"] === "string") return vars[o["$var"]];
    const out: Record<string, unknown> = {};
    for (const [k, y] of Object.entries(o)) out[k] = walkV(y);
    return out;
  };
  return walkV(v) as Record<string, unknown>;
}

File: kotlin/rayfold-core/src/main/kotlin/dev/rayfold/core/Cost.kt
package dev.rayfold.core

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

data class CostEstimate(val cost: Long, val depth: Int, val fields: Int)

/** Saturating Long arithmetic: a cost may grow without bound but never wraps around to a small or negative number. */
object Sat {
    fun add(a: Long, b: Long): Long {
        val r = a + b
        return if (((a xor r) and (b xor r)) < 0) (if (a < 0) Long.MIN_VALUE else Long.MAX_VALUE) else r
    }

    fun mul(a: Long, b: Long): Long = try {
        Math.multiplyExact(a, b)
    } catch (e: ArithmeticException) {
        if ((a < 0) != (b < 0)) Long.MIN_VALUE else Long.MAX_VALUE
    }
}

/**
 * Static cost (mirrors packages/server/src/cost.ts): op base + perItem*first + per-field costs multiplied by enclosing
 * page sizes. Bad input can only raise an estimate: a page size that is not a whole number in [0, 200] (negative,
 * huge, fractional, a string, a `$ref`, a missing variable) counts as 200, every op costs at least 1, and the
 * arithmetic saturates. Planning passes coerced op args when it has them, so a valid page size is already clamped.
 */
class Cost(private val ir: RayfoldSchemaIR, private val views: Views) {
    private class Acc(var fields: Int = 0, var depth: Int = 0)

    fun estimate(op: OpDef, args: JsonObject, shape: Shape, vars: JsonObject): CostEstimate {
        val c = op.annotations.find("cost")
        val base = weight(c?.args?.get("base"), 1)
        val perItem = weight(c?.args?.get("perItem"), if (op.returns.isPage) 1L else 0L)
        val first = if (op.returns.isPage) pageFirst(args, op.args) else 1L
        val acc = Acc()
        val shapeCost = walk(op.returns, shape, 1, first, 1, acc, vars)
        return CostEstimate(maxOf(1L, Sat.add(Sat.add(base, Sat.mul(perItem, first)), shapeCost)), acc.depth, acc.fields)
    }

    private fun walk(t: TypeRef, shape: Shape, mult: Long, itemsFirst: Long, depth: Int, acc: Acc, vars: JsonObject): Long {
        acc.depth = maxOf(acc.depth, depth)
        val isPage = t.isPage || (t.isList && t.of?.isPage == true)
        var total = 0L
        fun visit(items: List<ShapeItem>, ref: TypeRef) {
            val union = ir.types[ref.baseName()]
            if (union?.kind == "union") {
                // every member answers a union's bare fields, spreads and deferred items, so the dearest member counts;
                // looked up on the union itself they cost nothing (mirrors cost.ts)
                val bare = items.filter { it.kind != "on" }
                val total0 = total
                val fields0 = acc.fields
                var worst = 0L
                var worstFields = 0
                for (m in union.members) {
                    total = 0L
                    acc.fields = 0
                    visit(bare, TypeRef("named", m))
                    worst = maxOf(worst, total)
                    worstFields = maxOf(worstFields, acc.fields)
                }
                total = Sat.add(total0, worst)
                acc.fields = fields0 + worstFields
                for (it in items) if (it.kind == "on") it.shape?.let { s -> visit(s.items, TypeRef("named", it.type)) }
                return
            }
            val fields = ir.fieldsOf(ref) ?: emptyList()
            for (it in items) when (it.kind) {
                "field" -> {
                    acc.fields++
                    val f = fields.firstOrNull { x -> x.name == it.name }
                    val fc = f?.annotations?.find("cost")
                    val childFirst = if (f != null && f.type.isPage) {
                        pageFirst(substituteVars(JsonObject(it.args ?: emptyMap()), vars) as? JsonObject ?: JsonObject(emptyMap()), f.args)
                    } else 1L
                    // scalars and enums come with the row already loaded; a field that returns objects costs 1, and a page
                    // 1 more per row, unless @cost says otherwise
                    val defaultBase = if (f != null && ir.isScalarLike(f.type)) 0L else 1L
                    val defaultPerItem = if (f != null && f.type.isPage) 1L else 0L
                    val own = Sat.add(weight(fc?.args?.get("base"), defaultBase), Sat.mul(weight(fc?.args?.get("perItem"), defaultPerItem), childFirst))
                    total = Sat.add(total, Sat.mul(mult, own))
                    if (f != null && !ir.isScalarLike(f.type)) {
                        val sub = it.shape ?: views.defaultShape(f.type)
                        val childMult = if (isPage && it.name == "items") Sat.mul(mult, itemsFirst) else mult
                        total = Sat.add(total, walk(f.type, sub, childMult, childFirst, depth + 1, acc, vars))
                    }
                }
                "spread" -> ir.views["${it.type}.${it.view}"]?.let { v -> visit(v.shape.items, ref) }
                "on" -> it.shape?.let { s -> visit(s.items, TypeRef("named", it.type)) }
                "defer" -> it.shape?.let { s -> visit(s.items, ref) }
            }
        }
        visit(shape.items, t)
        return total
    }

    private fun pageFirst(args: JsonObject, defs: List<ArgDef>): Long {
        // the PageArgs argument by its type, whatever the schema calls it (mirrors cost.ts)
        val name = defs.firstOrNull { it.type.baseName() == "PageArgs" }?.name ?: "page"
        when (val page = args[name]) {
            null, JsonNull -> {}
            is JsonObject -> {
                if (Args.isRef(page)) return MAX_PAGE
                page["first"]?.let { return pageSize(it) }
            }
            else -> return MAX_PAGE
        }
        args["first"]?.let { return pageSize(it) }
        (defs.firstOrNull { it.name == name }?.default as? JsonObject)?.get("first")?.let { return pageSize(it) }
        defs.firstOrNull { it.name == "first" }?.default?.let { return pageSize(it) }
        return 20
    }

    private fun pageSize(v: JsonElement): Long = StrictJson.integerOrNull(v)?.toLongOrNull()?.takeIf { it in 0..MAX_PAGE } ?: MAX_PAGE

    /** A schema `@cost` weight; never negative, so an annotation cannot lower what the shape adds up to. */
    private fun weight(v: JsonElement?, default: Long): Long = v.numberOrNull()?.takeIf { it.isFinite() }?.let { maxOf(0.0, it).toLong() } ?: default

    companion object {
        private val MAX_PAGE = Args.MAX_PAGE_FIRST.toLong()

        fun substituteVars(v: JsonElement, vars: JsonObject): JsonElement = when (v) {
            is JsonArray -> JsonArray(v.map { substituteVars(it, vars) })
            is JsonObject -> {
                val name = (v["\$var"] as? JsonPrimitive)?.takeIf { it.isString }?.content
                // lenient: estimation runs before execution; the executor reports a missing variable
                if (name != null && v.size == 1) vars[name] ?: JsonNull
                else JsonObject(v.mapValues { (_, x) -> substituteVars(x, vars) })
            }
            else -> v
        }
    }
}
